File: tom_som_dart_runtime/lib/src/spec_meta_bridge.dart
/// Bridge from the meta-JSON class graph ([SpecModel]) to the canonical
/// metadata tree ([SomMetaTree], DR1 §3.1).
///
/// The generated facades (DR8) will emit populated [SomMetaTree]s directly;
/// until then, every consumer that only has the exported spec-model meta-data
/// (the conformance harness, the clitool, the editor) builds its tree here.
/// The expansion follows the same walk [SpecReflection.resolve] performs —
/// crucially, a node's [SomMetaNode.sectionId] is the **field-level** id
/// exactly as exported (`field.sectionId`), so the tree's path grammar stays
/// byte-compatible with the paths a [SpecDocument] is keyed by. DR1 §2.2's
/// "field id, else target-class id" resolution for a section/complex node's
/// display *key* is carried alongside it in [SomMetaNode.classSectionId] (the
/// target class's own id) — a codec renders `sectionId ?? classSectionId`
/// without re-consulting the model, and neither id enters [SomMetaNode.segment].
///
/// Recursive class references become terminal re-entry nodes
/// ([SomMetaNode.recursive]), mirroring how the generated chains terminate.
library;

import 'spec_meta.dart';
import 'spec_model.dart';

/// Annotation names that have a dedicated [SomMetaNode] slot; everything else
/// is captured losslessly into [SomMetaNode.extra].
const Set<String> _slottedAnnotations = {
  'SectionId',
  'SectionIdPattern',
  'SerializationOrder',
  'Min',
  'Unused',
  'ContentType',
  'ContentHelp',
  'Headline',
  'Comment',
  'Form',
  'Document',
  'MapsTo',
  'DetailedIn',
  'SecondLevelSectionId',
};

/// Builds the wired [SomMetaTree] for one document root of [model].
///
/// When [rootType] is omitted the model's first root is used; otherwise the
/// root is looked up via [SpecModel.rootByType] (throwing [ArgumentError] for
/// an unknown type). Children are ordered by `@SerializationOrder` (annotated
/// members first, then declaration order), which is the sibling emission order
/// of the hierarchical YAML format (DR1 §2.3).
SomMetaTree buildSomMetaTree(SpecModel model, {String? rootType}) {
  final root = rootType == null && model.roots.isNotEmpty
      ? model.roots.first
      : model.rootByType(rootType ?? '');
  final cls = model.classNamed(root.type);
  final rootNode = SomMetaNode(
    className: root.type,
    sectionId: root.sectionId ?? cls?.sectionId,
    classSectionId: cls?.sectionId,
    kind: SomMetaKind.section,
    typeName: root.type,
    headline: cls?.headline,
    docComment: root.doc ?? cls?.doc,
    classDocComment: cls?.doc,
    mapsTo: cls?.mapsTo,
    detailedIn: cls?.detailedIn,
    document: SomDocMeta(
      name: root.title,
      description: root.description ?? '',
      basedOn: _basedOn(cls),
    ),
    secondLevelIds: _secondLevelIds(cls?.annotations ?? const []),
    extra: _extras(cls?.annotations ?? const []),
    children: cls == null
        ? const []
        : _childNodes(model, cls, <String>{root.type}),
  );
  return SomMetaTree(rootNode);
}

List<String> _basedOn(SpecClass? cls) {
  final raw = cls?.annotation('Document')?.argument('basedOn');
  if (raw is List) return raw.map((e) => '$e').toList();
  return const [];
}

List<SomMetaNode> _childNodes(
    SpecModel model, SpecClass cls, Set<String> stack) {
  final indexed = cls.fields.asMap().entries.toList()
    ..sort((a, b) {
      const fallback = 1 << 30;
      final oa = a.value.serializationOrder ?? fallback;
      final ob = b.value.serializationOrder ?? fallback;
      return oa != ob ? oa.compareTo(ob) : a.key.compareTo(b.key);
    });
  return [
    for (final e in indexed) _fieldNode(model, cls, e.value, stack),
  ];
}

SomMetaNode _fieldNode(
    SpecModel model, SpecClass owner, SpecField field, Set<String> stack) {
  final kind = _kindOf(field.kind);
  SpecClass? target;
  var recursive = false;
  List<SomMetaNode> children = const [];
  SomMetaNode? elementNode;

  if (field.kind == SpecFieldKind.complex ||
      field.kind == SpecFieldKind.section) {
    final typeName = field.type;
    target = model.classNamed(typeName);
    if (target != null) {
      if (stack.contains(target.name)) {
        recursive = true;
      } else {
        stack.add(target.name);
        children = _childNodes(model, target, stack);
        stack.remove(target.name);
      }
    }
  } else if (field.kind == SpecFieldKind.list && field.elementIsComplex) {
    final element = model.classNamed(field.elementType);
    if (element != null) {
      List<SomMetaNode> elementChildren = const [];
      var elementRecursive = false;
      if (stack.contains(element.name)) {
        elementRecursive = true;
      } else {
        stack.add(element.name);
        elementChildren = _childNodes(model, element, stack);
        stack.remove(element.name);
      }
      elementNode = SomMetaNode(
        className: element.name,
        classSectionId: element.sectionId,
        kind: SomMetaKind.complex,
        typeName: element.name,
        headline: element.headline,
        docComment: element.doc,
        classDocComment: element.doc,
        mapsTo: element.mapsTo,
        detailedIn: element.detailedIn,
        recursive: elementRecursive,
        children: elementChildren,
      );
    }
  }

  return SomMetaNode(
    className: target?.name ?? owner.name,
    memberName: field.name,
    sectionId: field.sectionId,
    classSectionId: target?.sectionId,
    sectionIdPattern: field.sectionIdPattern,
    kind: kind,
    typeName: field.type ?? field.elementType ?? field.enumType ?? 'String',
    serializationOrder: field.serializationOrder,
    min: field.min,
    unused: field.annotation('Unused') != null,
    contentType: field.contentType == null
        ? null
        : SomContentTypeMeta(
            type: field.contentType!,
            description:
                '${field.annotation('ContentType')?.argument('description') ?? ''}',
          ),
    contentHelp: field.help,
    headline: field.headline ?? target?.headline,
    comment: '${field.annotation('Comment')?.argument('text') ?? ''}'.isEmpty
        ? null
        : '${field.annotation('Comment')!.argument('text')}',
    docComment: field.doc ?? target?.doc,
    classDocComment: target?.doc,
    form: field.kind == SpecFieldKind.form
        ? SomFormMeta(fields: [
            for (var i = 0; i < field.formFields.length; i++)
              SomFormFieldMeta(
                name: field.formFields[i].name,
                typeName: field.formFields[i].type,
                description: field.formFields[i].label,
                required: field.formFields[i].required,
                hint: field.formFields[i].hint,
                order: i,
                enumValues: field.formFields[i].enumValues,
              ),
          ])
        : null,
    mapsTo: target?.mapsTo,
    detailedIn: target?.detailedIn,
    secondLevelIds: _secondLevelIds(field.annotations),
    extra: _extras(field.annotations),
    recursive: recursive,
    children: children,
    elementNode: elementNode,
  );
}

SomMetaKind _kindOf(SpecFieldKind kind) {
  switch (kind) {
    case SpecFieldKind.list:
      return SomMetaKind.list;
    case SpecFieldKind.form:
      return SomMetaKind.form;
    case SpecFieldKind.section:
      return SomMetaKind.section;
    case SpecFieldKind.content:
      return SomMetaKind.content;
    case SpecFieldKind.enumValue:
      return SomMetaKind.enumValue;
    case SpecFieldKind.complex:
      return SomMetaKind.complex;
    case SpecFieldKind.scalar:
      return SomMetaKind.scalar;
  }
}

List<SomSecondLevelId> _secondLevelIds(List<SpecAnnotation> annotations) => [
      for (final a in annotations)
        if (a.name == 'SecondLevelSectionId')
          SomSecondLevelId(
            documentClass: '${a.argument('documentClass') ?? ''}',
            id: '${a.argument('id') ?? ''}',
          ),
    ];

List<SomMetaExtra> _extras(List<SpecAnnotation> annotations) => [
      for (final a in annotations)
        if (!_slottedAnnotations.contains(a.name))
          SomMetaExtra(annotation: a.name, args: a.arguments),
    ];

File: tom_som_javascript_runtime/tom_som_runtime/spec_meta_bridge.js
'use strict';

/**
 * Bridge from the meta-JSON class graph ({@link SpecModel}) to the canonical
 * metadata tree ({@link SomMetaTree}, SOM §7.1) — a faithful port of
 * `tom_som_dart_runtime/lib/src/spec_meta_bridge.dart` (and
 * `spec_meta_bridge.py`).
 *
 * The generated facades (SOM §8) emit populated {@link SomMetaTree}s directly;
 * every consumer that only has the exported spec-model meta-data (the
 * conformance harness, tooling) builds its tree here. The expansion follows
 * the same walk {@link SpecReflection#resolve} performs — crucially, a node's
 * {@link SomMetaNode#sectionId} is the **field-level** id exactly as exported
 * (`field.sectionId`), so the tree's path grammar stays byte-compatible with
 * the paths a {@link SpecDocument} is keyed by. (SOM §12.2's "field id, else
 * target-class id" resolution is applied at *export* time by the model
 * exporter / SOM generator, not re-derived here.)
 *
 * Recursive class references become terminal re-entry nodes
 * ({@link SomMetaNode#recursive}), mirroring how the generated chains
 * terminate.
 */

const {
  SomContentTypeMeta,
  SomDocMeta,
  SomFormFieldMeta,
  SomFormMeta,
  SomMetaExtra,
  SomMetaKind,
  SomMetaNode,
  SomMetaTree,
} = require('./spec_meta');
const { SpecFieldKind } = require('./spec_model');

/**
 * Annotation names that have a dedicated {@link SomMetaNode} slot; everything
 * else is captured losslessly into {@link SomMetaNode#extra}.
 */
const _SLOTTED_ANNOTATIONS = new Set([
  'SectionId',
  'SectionIdPattern',
  'SerializationOrder',
  'Min',
  'Unused',
  'ContentType',
  'ContentHelp',
  'Headline',
  'Comment',
  'Form',
  'Document',
  'MapsTo',
  'DetailedIn',
]);

const _KIND_OF = Object.freeze({
  [SpecFieldKind.LIST]: SomMetaKind.LIST,
  [SpecFieldKind.FORM]: SomMetaKind.FORM,
  [SpecFieldKind.SECTION]: SomMetaKind.SECTION,
  [SpecFieldKind.CONTENT]: SomMetaKind.CONTENT,
  [SpecFieldKind.ENUM]: SomMetaKind.ENUM_VALUE,
  [SpecFieldKind.COMPLEX]: SomMetaKind.COMPLEX,
  [SpecFieldKind.SCALAR]: SomMetaKind.SCALAR,
});

/**
 * Builds the wired {@link SomMetaTree} for one document root of `model`.
 *
 * When `rootType` is omitted the model's first root is used; otherwise the
 * root is looked up via {@link SpecModel#rootByType} (throwing for an unknown
 * type). Children are ordered by `@SerializationOrder` (annotated members
 * first, then declaration order), which is the sibling emission order of the
 * hierarchical YAML format (SOM §12.3).
 *
 * @param {import('./spec_model').SpecModel} model
 * @param {string|null} [rootType]
 * @returns {SomMetaTree}
 */
function buildSomMetaTree(model, rootType = null) {
  const root =
    rootType === null || rootType === undefined
      ? model.roots.length > 0
        ? model.roots[0]
        : model.rootByType('')
      : model.rootByType(rootType);
  const cls = model.classNamed(root.type);
  const rootNode = new SomMetaNode({
    className: root.type,
    sectionId:
      root.sectionId != null ? root.sectionId : cls !== null ? cls.sectionId : null,
    classSectionId: cls !== null ? cls.sectionId : null,
    kind: SomMetaKind.SECTION,
    typeName: root.type,
    headline: cls !== null ? cls.headline : null,
    docComment: root.doc != null ? root.doc : cls !== null ? cls.doc : null,
    classDocComment: cls !== null ? cls.doc : null,
    mapsTo: cls !== null ? cls.mapsTo : null,
    detailedIn: cls !== null ? cls.detailedIn : null,
    document: new SomDocMeta({
      name: root.title,
      description: root.description != null ? root.description : '',
      basedOn: _basedOn(cls),
    }),
    extra: _extras(cls !== null ? cls.annotations : []),
    children: cls === null ? [] : _childNodes(model, cls, new Set([root.type])),
  });
  return new SomMetaTree(rootNode);
}

function _basedOn(cls) {
  const ann = cls !== null ? cls.annotation('Document') : null;
  const raw = ann !== null ? ann.argument('basedOn') : null;
  if (Array.isArray(raw)) {
    return raw.map((e) => String(e));
  }
  return [];
}

function _childNodes(model, cls, stack) {
  const fallback = 1 << 30;
  const indexed = cls.fields.map((f, i) => [i, f]);
  indexed.sort((a, b) => {
    const oa = a[1].serializationOrder !== null ? a[1].serializationOrder : fallback;
    const ob = b[1].serializationOrder !== null ? b[1].serializationOrder : fallback;
    if (oa !== ob) {
      return oa - ob;
    }
    return a[0] - b[0];
  });
  return indexed.map(([, f]) => _fieldNode(model, cls, f, stack));
}

function _fieldNode(model, owner, field, stack) {
  const kind = _KIND_OF[field.kind];
  let target = null;
  let recursive = false;
  let children = [];
  let elementNode = null;

  if (field.kind === SpecFieldKind.COMPLEX || field.kind === SpecFieldKind.SECTION) {
    target = model.classNamed(field.type);
    if (target !== null) {
      if (stack.has(target.name)) {
        recursive = true;
      } else {
        stack.add(target.name);
        children = _childNodes(model, target, stack);
        stack.delete(target.name);
      }
    }
  } else if (field.kind === SpecFieldKind.LIST && field.elementIsComplex) {
    const element = model.classNamed(field.elementType);
    if (element !== null) {
      let elementChildren = [];
      let elementRecursive = false;
      if (stack.has(element.name)) {
        elementRecursive = true;
      } else {
        stack.add(element.name);
        elementChildren = _childNodes(model, element, stack);
        stack.delete(element.name);
      }
      elementNode = new SomMetaNode({
        className: element.name,
        classSectionId: element.sectionId,
        kind: SomMetaKind.COMPLEX,
        typeName: element.name,
        headline: element.headline,
        docComment: element.doc,
        classDocComment: element.doc,
        mapsTo: element.mapsTo,
        detailedIn: element.detailedIn,
        recursive: elementRecursive,
        children: elementChildren,
      });
    }
  }

  let contentType = null;
  if (field.contentType !== null && field.contentType !== undefined) {
    const ctAnn = field.annotation('ContentType');
    contentType = new SomContentTypeMeta(
      field.contentType,
      String((ctAnn !== null ? ctAnn.argument('description') : null) || ''),
    );
  }

  const commentAnn = field.annotation('Comment');
  const commentText = String(
    (commentAnn !== null ? commentAnn.argument('text') : null) || '',
  );

  let form = null;
  if (field.kind === SpecFieldKind.FORM) {
    form = new SomFormMeta(
      field.formFields.map(
        (ff, i) =>
          new SomFormFieldMeta({
            name: ff.name,
            typeName: ff.type,
            description: ff.label,
            required: ff.required,
            hint: ff.hint,
            order: i,
          }),
      ),
    );
  }

  return new SomMetaNode({
    className: target !== null ? target.name : owner.name,
    memberName: field.name,
    sectionId: field.sectionId,
    classSectionId: target !== null ? target.sectionId : null,
    sectionIdPattern: field.sectionIdPattern,
    kind,
    typeName: field.type || field.elementType || field.enumType || 'String',
    serializationOrder: field.serializationOrder,
    min: field.min,
    unused: field.annotation('Unused') !== null,
    contentType,
    contentHelp: field.help,
    headline:
      field.headline != null
        ? field.headline
        : target !== null
          ? target.headline
          : null,
    comment: commentText !== '' ? commentText : null,
    docComment: field.doc || (target !== null ? target.doc : null),
    classDocComment: target !== null ? target.doc : null,
    form,
    mapsTo: target !== null ? target.mapsTo : null,
    detailedIn: target !== null ? target.detailedIn : null,
    extra: _extras(field.annotations),
    recursive,
    children,
    elementNode,
  });
}

function _extras(annotations) {
  return annotations
    .filter((a) => !_SLOTTED_ANNOTATIONS.has(a.name))
    .map((a) => new SomMetaExtra(a.name, a.arguments));
}

module.exports = { buildSomMetaTree };
